samples/tls_between_host_enclave/client/client.cpp

```diff
@@ -0,0 +1,280 @@
+            errno);
+        close(sockfd);
+        sockfd = -1;
+        goto done;
+    }
+    printf(TLS_CLIENT "connected to %s:%s\n", server_name, server_port);
+done:
+    return sockfd;
+}
+
+int main(int argc, char** argv)
+{
+    int ret = 1;
+    X509* cert = NULL;
+    SSL_CTX* ctx = NULL;
+    SSL* ssl = NULL;
+    int serversocket = 0;
+    char* server_name = NULL;
+    char* server_port = NULL;
+    int error = 0;
+
+    printf("\nStarting" TLS_CLIENT "\n\n\n");
+    if ((error = parse_arguments(argc, argv, &server_name, &server_port)) != 0)
+    {
+        printf(
+            TLS_CLIENT "TLS client:parse input parmeter failed (%d)!\n", error);
+        goto done;
+    }
+
+    // initialize openssl library and register algorithms
+    OpenSSL_add_all_algorithms();
+    ERR_load_BIO_strings();
+    ERR_load_crypto_strings();
+    SSL_load_error_strings();
+
+    if (SSL_library_init() < 0)
+    {
+        printf(TLS_CLIENT
+               "TLS client: could not initialize the OpenSSL library !\n");
+        goto done;
+    }
+
+    if ((ctx = SSL_CTX_new(TLS_client_method())) == NULL)
+    {
+        printf(TLS_CLIENT "TLS client: unable to create a new SSL context\n");
+        goto done;
+    }
+
+    // choose TLSv1.2 by excluding SSLv2, SSLv3 ,TLS 1.0 and TLS 1.1
+    SSL_CTX_set_options(ctx, SSL_OP_NO_SSLv2);
+    SSL_CTX_set_options(ctx, SSL_OP_NO_SSLv3);
+    SSL_CTX_set_options(ctx, SSL_OP_NO_TLSv1);
+    SSL_CTX_set_options(ctx, SSL_OP_NO_TLSv1_1);
+
+    // specify the verify_callback for custom verification
+    SSL_CTX_set_verify(ctx, SSL_VERIFY_PEER, &verify_callback);
+
+    if ((ssl = SSL_new(ctx)) == NULL)
+    {
+        printf(TLS_CLIENT
+               "Unable to create a new SSL connection state object\n");
+        goto done;
+    }
+
+    serversocket = create_socket(server_name, server_port);
+    if (serversocket == -1)
+    {
+        printf(
+            TLS_CLIENT
+            "create a socket and initate a TCP connect to server: %s:%s "
+            "(errno=%d)\n",
+            server_name,
+            server_port,
+            errno);
+        goto done;
+    }
+
+    // setup ssl socket and initiate TLS connection with TLS server
+    SSL_set_fd(ssl, serversocket);
+    if ((error = SSL_connect(ssl)) != 1)
+    {
+        printf(
+            TLS_CLIENT "Error: Could not establish an SSL session ret2=%d "
+                       "SSL_get_error()=%d\n",
+            error,
+            SSL_get_error(ssl, error));
+        goto done;
+    }
+    printf(
+        TLS_CLIENT "successfully established TLS channel:%s\n",
+        SSL_get_version(ssl));
+
+    // start the client server communication
+    if ((error = communicate_with_server(ssl)) != 0)
+    {
+        printf(TLS_CLIENT "Failed: communicate_with_server (ret=%d)\n", error);
+        goto done;
+    }
+
+    // Free the structures we don't need anymore
+    if (serversocket != -1)
+        close(serversocket);
+
+    SSL_free(ssl);
```

Comment:
can move frees to after done if null checks are added

Reply:
good idea, moved